tests/kernel/wave/common/utils.py

```diff
@@ -37,3 +40,40 @@ def param_bool(name, shortname=None, values=None):
     values = values or [False, True]
     ids = [f"{shortname}" if v else f"no_{shortname}" for v in values]
     return pytest.mark.parametrize(name, [pytest.param(v) for v in values], ids=ids)
+
+
+def scaled_dot_product_attention_bhsd(
+    query: Tensor, key: Tensor, value: Tensor, is_causal: bool = False
+) -> Tensor:
+    """
+    This version mimics PyTorch's `torch.nn.functional.scaled_dot_product_attention`
+    with optional causal masking and improved numerical stability.
+    Intended for comparison and debugging purposes.
+
+    Args:
+        query (Tensor): query tensor of shape [B, H, S_q, D].
+        key (Tensor): key tensor of shape [B, H, S_k, D].
+        value (Tensor): value tensor of shape [B, H, S_k, D].
+        is_causal (bool): If True, applies causal masking to the attention logits.
+
+    Returns:
+        Tensor: Output tensor of shape [B, H, S_q, D] after applying attention.
+    """
+    scale: float = query.shape[-1] ** -0.5
+    attn_logits: Tensor = torch.matmul(query, key.transpose(-2, -1)) * scale
+
```

Comment:
as discussed offline, let's try use f32 softmax, maybe it would be closer to our wave output :)

Reply:
I have added checks for the `dtype `and `update to fp32` accordingly.